File: examples/python/agent_openai.py
#!/usr/bin/env python3
"""
NCHE Agent Example - OpenAI GPT

This example demonstrates how to build an AI agent using OpenAI's GPT
that executes actions through NCHE for human oversight.

Requirements:
    pip install openai requests

Usage:
    export OPENAI_API_KEY=your_api_key
    export NCHE_API_KEY=your_nche_api_key
    python agent_openai.py
"""

import json
import os
from openai import OpenAI
from nche_client import NcheClient

# Configuration
NCHE_URL = os.getenv("NCHE_URL", "http://localhost:3000")
NCHE_API_KEY = os.getenv("NCHE_API_KEY")
OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")

# Define tools that the agent can use (these map to NCHE tools)
TOOLS = [
    {
        "type": "function",
        "function": {
            "name": "send_email",
            "description": "Send an email to a recipient. Use this when you need to communicate via email.",
            "parameters": {
                "type": "object",
                "properties": {
                    "to": {
                        "type": "string",
                        "description": "Email address of the recipient",
                    },
                    "subject": {
                        "type": "string",
                        "description": "Subject line of the email",
                    },
                    "body": {
                        "type": "string",
                        "description": "Body content of the email",
                    },
                },
                "required": ["to", "subject", "body"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "http_request",
            "description": "Make an HTTP request to an API endpoint.",
            "parameters": {
                "type": "object",
                "properties": {
                    "method": {
                        "type": "string",
                        "enum": ["GET", "POST", "PUT", "DELETE"],
                        "description": "HTTP method",
                    },
                    "url": {
                        "type": "string",
                        "description": "URL to request",
                    },
                    "headers": {
                        "type": "object",
                        "description": "HTTP headers",
                    },
                    "body": {
                        "type": "string",
                        "description": "Request body (for POST/PUT)",
                    },
                },
                "required": ["method", "url"],
            },
        },
    },
]


def execute_tool_via_nche(nche: NcheClient, tool_name: str, tool_input: dict) -> str:
    """Execute a tool through NCHE and return the result."""
    print(f"\n[NCHE] Proposing action: {tool_name}")
    print(f"[NCHE] Parameters: {json.dumps(tool_input, indent=2)}")

    action = nche.execute_tool(tool_name, tool_input, wait=True, timeout=120)

    print(f"[NCHE] Action state: {action.state}")

    if action.state == "executed":
        return json.dumps(action.result or {"status": "success"})
    elif action.state == "denied":
        return json.dumps({"error": "Action was denied by human reviewer"})
    elif action.state == "failed":
        return json.dumps({"error": action.error or "Action execution failed"})
    else:
        return json.dumps({"error": f"Unexpected state: {action.state}"})


def run_agent(user_message: str):
    """Run the agent with a user message."""
    if not NCHE_API_KEY:
        raise ValueError("NCHE_API_KEY environment variable is required")
    if not OPENAI_API_KEY:
        raise ValueError("OPENAI_API_KEY environment variable is required")

    # Initialize clients
    nche = NcheClient(NCHE_URL, NCHE_API_KEY)
    openai = OpenAI(api_key=OPENAI_API_KEY)

    # Create NCHE session
    print("[NCHE] Creating session...")
    session = nche.create_session(
        actor_id="gpt-agent",
        actor_type="agent",
        autonomy_level="supervised",  # Require human approval
    )
    print(f"[NCHE] Session created: {session.id}")

    try:
        # Initialize conversation
        messages = [
            {
                "role": "system",
                "content": """You are a helpful assistant that can send emails and make HTTP requests.
When you need to perform an action, use the provided tools.
All actions are reviewed by humans before execution for safety.""",
            },
            {"role": "user", "content": user_message},
        ]

        print(f"\n[User] {user_message}\n")

        # Agent loop
        while True:
            # Call GPT
            response = openai.chat.completions.create(
                model="gpt-4o",
                messages=messages,
                tools=TOOLS,
                tool_choice="auto",
            )

            message = response.choices[0].message

            # Check if we're done (no tool calls)
            if not message.tool_calls:
                if message.content:
                    print(f"[GPT] {message.content}")
                break

            # Print any text content
            if message.content:
                print(f"[GPT] {message.content}")

            # Add assistant message to conversation
            messages.append(message)

            # Execute tools through NCHE
            for tool_call in message.tool_calls:
                tool_name = tool_call.function.name
                tool_input = json.loads(tool_call.function.arguments)

                result = execute_tool_via_nche(nche, tool_name, tool_input)

                # Add tool result to conversation
                messages.append(
                    {
                        "role": "tool",
                        "tool_call_id": tool_call.id,
                        "content": result,
                    }
                )

    finally:
        # Clean up session
        print("\n[NCHE] Ending session...")
        nche.end_session()
        print("[NCHE] Session ended")


if __name__ == "__main__":
    # Example: Ask the agent to send an email
    run_agent(
        "Please send an email to team@example.com with subject 'Weekly Update' "
        "and body 'Hi team, here is this week's progress update. All tasks are on track.'"
    )

File: examples/rust/src/agent_openai.rs
//! NCHE Agent Example - OpenAI GPT
//!
//! This example demonstrates how to build an AI agent using OpenAI's GPT
//! that executes actions through NCHE for human oversight.
//!
//! # Usage
//! ```bash
//! export OPENAI_API_KEY=your_api_key
//! export NCHE_API_KEY=your_nche_api_key
//! cargo run --bin agent_openai
//! ```

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::env;
use std::time::Duration;

mod nche_client;
use nche_client::NcheClient;

// OpenAI API types
#[derive(Debug, Serialize)]
struct OpenAIRequest {
    model: String,
    messages: Vec<OpenAIMessage>,
    tools: Vec<OpenAITool>,
    tool_choice: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
struct OpenAIMessage {
    role: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    content: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    tool_calls: Option<Vec<ToolCall>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    tool_call_id: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
struct ToolCall {
    id: String,
    #[serde(rename = "type")]
    call_type: String,
    function: FunctionCall,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
struct FunctionCall {
    name: String,
    arguments: String,
}

#[derive(Debug, Serialize)]
struct OpenAITool {
    #[serde(rename = "type")]
    tool_type: String,
    function: OpenAIFunction,
}

#[derive(Debug, Serialize)]
struct OpenAIFunction {
    name: String,
    description: String,
    parameters: Value,
}

#[derive(Debug, Deserialize)]
struct OpenAIResponse {
    choices: Vec<Choice>,
}

#[derive(Debug, Deserialize)]
struct Choice {
    message: OpenAIMessage,
}

fn get_tools() -> Vec<OpenAITool> {
    vec![
        OpenAITool {
            tool_type: "function".to_string(),
            function: OpenAIFunction {
                name: "send_email".to_string(),
                description: "Send an email to a recipient.".to_string(),
                parameters: json!({
                    "type": "object",
                    "properties": {
                        "to": { "type": "string", "description": "Email address" },
                        "subject": { "type": "string", "description": "Subject line" },
                        "body": { "type": "string", "description": "Email body" }
                    },
                    "required": ["to", "subject", "body"]
                }),
            },
        },
        OpenAITool {
            tool_type: "function".to_string(),
            function: OpenAIFunction {
                name: "http_request".to_string(),
                description: "Make an HTTP request.".to_string(),
                parameters: json!({
                    "type": "object",
                    "properties": {
                        "method": { "type": "string", "enum": ["GET", "POST", "PUT", "DELETE"] },
                        "url": { "type": "string", "description": "URL to request" },
                        "headers": { "type": "object" },
                        "body": { "type": "string" }
                    },
                    "required": ["method", "url"]
                }),
            },
        },
    ]
}

async fn call_openai(
    client: &reqwest::Client,
    api_key: &str,
    messages: &[OpenAIMessage],
) -> Result<OpenAIResponse> {
    let request = OpenAIRequest {
        model: "gpt-4o".to_string(),
        messages: messages.to_vec(),
        tools: get_tools(),
        tool_choice: "auto".to_string(),
    };

    let response = client
        .post("https://api.openai.com/v1/chat/completions")
        .header("Authorization", format!("Bearer {}", api_key))
        .header("Content-Type", "application/json")
        .json(&request)
        .send()
        .await
        .context("Failed to call OpenAI API")?;

    if !response.status().is_success() {
        let error = response.text().await?;
        anyhow::bail!("OpenAI API error: {}", error);
    }

    response.json().await.context("Failed to parse response")
}

async fn execute_tool_via_nche(
    nche: &NcheClient,
    session_id: &str,
    tool_name: &str,
    tool_input: Value,
) -> Result<String> {
    println!("\n[NCHE] Proposing action: {}", tool_name);
    println!("[NCHE] Parameters: {}", serde_json::to_string_pretty(&tool_input)?);

    let action = nche.propose_action(session_id, tool_name, tool_input).await?;
    let action = nche.wait_for_action(&action.id, Duration::from_secs(120)).await?;

    println!("[NCHE] Action state: {}", action.state);

    Ok(match action.state.as_str() {
        "executed" => serde_json::to_string(&action.result.unwrap_or(json!({"status": "success"})))?,
        "denied" => json!({"error": "Action was denied by human reviewer"}).to_string(),
        "failed" => json!({"error": action.error.unwrap_or_else(|| "Action failed".to_string())}).to_string(),
        state => json!({"error": format!("Unexpected state: {}", state)}).to_string(),
    })
}

#[tokio::main]
async fn main() -> Result<()> {
    let nche_url = env::var("NCHE_URL").unwrap_or_else(|_| "http://localhost:3000".to_string());
    let nche_api_key = env::var("NCHE_API_KEY").context("NCHE_API_KEY is required")?;
    let openai_api_key = env::var("OPENAI_API_KEY").context("OPENAI_API_KEY is required")?;

    let http_client = reqwest::Client::new();
    let nche = NcheClient::new(&nche_url, &nche_api_key);

    // Create NCHE session
    println!("[NCHE] Creating session...");
    let session = nche.create_session("gpt-agent", "agent", "supervised").await?;
    println!("[NCHE] Session created: {}", session.id);

    let user_message = "Please send an email to team@example.com with subject 'Weekly Update' \
                        and body 'Hi team, here is this week's progress update.'";

    println!("\n[User] {}\n", user_message);

    let mut messages = vec![
        OpenAIMessage {
            role: "system".to_string(),
            content: Some("You are a helpful assistant that can send emails and make HTTP requests. \
                          All actions are reviewed by humans before execution.".to_string()),
            tool_calls: None,
            tool_call_id: None,
        },
        OpenAIMessage {
            role: "user".to_string(),
            content: Some(user_message.to_string()),
            tool_calls: None,
            tool_call_id: None,
        },
    ];

    // Agent loop
    loop {
        let response = call_openai(&http_client, &openai_api_key, &messages).await?;
        let message = &response.choices[0].message;

        // Print any text content
        if let Some(content) = &message.content {
            if !content.is_empty() {
                println!("[GPT] {}", content);
            }
        }

        // Check if there are tool calls
        let tool_calls = message.tool_calls.clone().unwrap_or_default();
        if tool_calls.is_empty() {
            break;
        }

        // Add assistant message
        messages.push(message.clone());

        // Execute tools and add results
        for tool_call in tool_calls {
            let tool_input: Value = serde_json::from_str(&tool_call.function.arguments)
                .context("Failed to parse tool arguments")?;

            let result = execute_tool_via_nche(
                &nche,
                &session.id,
                &tool_call.function.name,
                tool_input,
            ).await?;

            messages.push(OpenAIMessage {
                role: "tool".to_string(),
                content: Some(result),
                tool_calls: None,
                tool_call_id: Some(tool_call.id),
            });
        }
    }

    // End session
    println!("\n[NCHE] Ending session...");
    nche.end_session(&session.id).await?;
    println!("[NCHE] Session ended");

    Ok(())
}
